producer/types.go

```diff
@@ -0,0 +1,152 @@
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package producer
+
+import (
+	"github.com/m3db/m3cluster/services"
+)
+
+// DataFinalizeReason defines the reason why the data is being finalized by Producer.
+type DataFinalizeReason int
+
+const (
+	// Consumed means the data has been fully consumed.
+	Consumed DataFinalizeReason = iota
+
+	// Dropped means the data has been dropped.
+	Dropped
+)
+
+// Data contains the data that will be produced by the producer.
+// It should only be finalized by the producer.
+type Data interface {
+	// Shard returns the shard of the data.
+	Shard() uint32
+
+	// Bytes returns the bytes of the data.
+	Bytes() []byte
+
+	// Size returns the size of the bytes of the data.
+	Size() uint32
+
+	// Finalize will be called by producer to indicate the end of its lifecycle.
+	Finalize(DataFinalizeReason)
+}
+
+// Producer produces data to a topic.
+type Producer interface {
+	// Produce produces the data.
+	Produce(data Data) error
+
+	// RegisterFilter registers a filter to a consumer service.
+	RegisterFilter(sid services.ServiceID, fn FilterFunc)
+
+	// RemoveFilter removes the filter of a consumer service.
+	RemoveFilter(sid services.ServiceID)
+
+	// Init initializes a producer.
+	Init()
+
+	// Close stops the producer from accepting new requests immediately.
+	// It will also block until all the data buffered in Producer has been consumed.
+	Close()
+}
+
+// FilterFunc can filter data.
+type FilterFunc func(data Data) bool
+
+// Options configs a producer.
+type Options interface {
+	// Buffer returns the buffer.
+	Buffer() Buffer
+
+	// SetBuffer sets the buffer.
+	SetBuffer(value Buffer) Options
+
+	// Writer returns the writer.
+	Writer() Writer
+
+	// SetWriter sets the writer.
+	SetWriter(value Writer) Options
+}
+
+// Buffer buffers all the data in the producer.
+type Buffer interface {
+	// Buffer buffers a data and returns a reference counted data.
+	Buffer(data Data) (RefCountedData, error)
+
+	// Init initializes the buffer.
+	Init()
+
+	// Close stops the buffer from accepting new requests immediately.
+	// It will also block until all the data buffered has been consumed.
+	Close()
+}
+
+// Writer writes all the data out to the consumer services.
+type Writer interface {
+	// Write writes a reference counted data out.
+	Write(shard uint32, d RefCountedData) error
+
+	// RegisterFilter registers a filter to a consumer service.
+	RegisterFilter(sid services.ServiceID, fn FilterFunc)
+
+	//RemoveFilter removes the filter of a consumer service.
+	RemoveFilter(sid services.ServiceID)
+
+	// Init initializes a writer.
+	Init() error
+
+	// Close closes the writer.
+	Close()
+}
+
+// RefCountedData is a reference counted data.
+type RefCountedData interface {
+	// IncRef increments the ref count.
```

Comment:
Might be better to pull `IncRef` and `DecRef` into a separate `RefCounted` interface.

Reply:
I'd prefer to keep it here since the IncRef/DecRef is not reused anywhere else and the lifecycle is managed by IncRef/DecRef and Drop together